### PR TITLE
swamp: initial structure of the tool

### PR DESCRIPTION
**Pull request by Bidon15**
Swamp with FullMeshLink network
First test with sync scenario

Relates to #262

### CODE REVIEW

File: node/tests/swamp/swamp.go
Comment (liamsi) [inline node/tests/swamp/swamp.go:30]: Maybe stupid question but why is this actually called swamp? Is this common in testing? my understanding is that this swamp is a mix of integration tests that leani a bit towards e2e tests.

File: node/tests/README.md
Comment (liamsi) [inline node/tests/README.md:1]: Maybe start with a very short description by having a simple diagram and a few bullet-points. It would also help reviewers as an intro to the PR.

File: node/tests/swamp/swamp.go
Comment (liamsi) [inline node/tests/swamp/swamp.go:63]: Was this forgotten or is it intentional to use fmt.Println here?

File: node/tests/swamp/sync_test.go
Comment (liamsi) [inline node/tests/swamp/sync_test.go:40]: Why 1 second?

File: node/tests/swamp/sync_test.go
Comment (liamsi) [inline node/tests/swamp/sync_test.go:51]: same

File: node/tests/swamp/swamp.go
Comment (adlerjohn) [reply node/tests/swamp/swamp.go:30]: `it's all ogre now`

Comment (Wondertan) [reply node/tests/swamp/swamp.go:30]: > it's all ogre now
:rofl:

File: node/tests/swamp/swamp.go
Comment (renaynay) [inline node/tests/swamp/swamp.go:39]: Can you make this a TODO with your username ?

File: node/tests/swamp/swamp.go
Comment (renaynay) [inline node/tests/swamp/swamp.go:72]:
```suggestion
// createPeer is a helper for celestia nodes to initialize
```

you don't need to add the `()` (this is true for all godocs)

File: node/tests/swamp/sync_test.go
Comment (renaynay) [inline node/tests/swamp/sync_test.go:40]:
```suggestion
	for {
		if !bridge.HeaderServ.IsSyncing() {
			break
		}
	}
```

File: node/tests/swamp/sync_test.go
Comment (renaynay) [inline node/tests/swamp/sync_test.go:10]: I don't like this dependency because the methods aren't documented / easy to follow. Can we please stick to assert / require for now?

**Review by renaynay — COMMENTED** [review]
Initial review

File: core/testing.go
Comment (Wondertan) [inline core/testing.go:26]: Look at him, sitting there and looking like nothing has happened. How dare you!

File: node/tests/swamp/sync_test.go
Comment (Wondertan) [inline node/tests/swamp/sync_test.go:27]: You don't need this

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:95]: maybe mention trusted or genesis in the name?

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:85]: Would be nice to mention where did you get this code from.

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:48]: ctx consitency?

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:30]: No it's not. However, it is common to have the naming to be bees related like `hive` or `x-swarm`, so I opted for some Shrek's swamp 😸

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:63]: This is forgotten after finished debugging session. Thanks for the finding!

File: node/tests/swamp/sync_test.go
Comment (Bidon15) [reply node/tests/swamp/sync_test.go:40]: It's related to the IsSyncing from service.go. We need to rework the method to be a subscription so I can utilize the asyncAssert functionality correctly.

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:39]: Removed

File: node/tests/swamp/sync_test.go
Comment (Bidon15) [reply node/tests/swamp/sync_test.go:40]: I'll leave the sleep and then redo once we have the new version of sync

File: node/tests/swamp/sync_test.go
Comment (Bidon15) [reply node/tests/swamp/sync_test.go:27]: Removed

File: core/testing.go
Comment (Bidon15) [reply core/testing.go:26]: Yeah. True. Both went through the storm to find this 😅

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:48]: Nice catch! Changed to ctx

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:85]: Agree. Added the comment section

File: node/tests/swamp/sync_test.go
Comment (Bidon15) [reply node/tests/swamp/sync_test.go:10]: Agreed. Removed for now

File: node/tests/README.md
Comment (Bidon15) [reply node/tests/README.md]: Done

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:94]: I would say trusted is more preferable. Changing

File: service/header/service.go
Comment (renaynay) [inline service/header/service.go:48]: I think we can just introduce a new method for this. `IsSyncing` is a valuable endpoint for RPC, so we should leave this one as is and implement a new method that will give you the functionality you need.

File: node/tests/README.md
Comment (adlerjohn) [inline node/tests/README.md:15]: all over, `celestia` -> `Celestia`

File: node/tests/README.md
Comment (adlerjohn) [inline node/tests/README.md:17]: Can you document what "linking" and "connecting" means?

File: node/tests/README.md
Comment (adlerjohn) [inline node/tests/README.md:20]: For consistency, use American English everywhere, e.g. initialization.

**Review by adlerjohn — COMMENTED** [review]
Some nits

File: core/testing.go
Comment (Bidon15) [reply core/testing.go:25]: I'll go 1 step further with this and separate creation of the kv from starting the mock node

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:30]: Sure. Thanks for pointing this

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:52]: On it

File: service/header/service.go
Comment (Bidon15) [reply service/header/service.go:48]: Ok. I will separate this todo from this method

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:109]: Got it

File: node/tests/README.md
Comment (Bidon15) [reply node/tests/README.md:15]: Changed

File: node/tests/README.md
Comment (Bidon15) [reply node/tests/README.md:20]: Ok

File: node/tests/swamp/swamp.go
Comment (adlerjohn) [inline node/tests/swamp/swamp.go:26]: Can you link to an exact commit and line on GitHub here? This documentation is brittle.

**Comment by Bidon15** [comment]
I've heavily messed up with pulling latest so I'm going to force push to eliminate some of the confusion 😓

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go]: I've made a separate issue #350 as the links look a lil bit ugly in the code comments section

File: service/header/service.go
Comment (renaynay) [inline service/header/service.go:49]: can you make this into an issue instead of a code comment?

File: Makefile
Comment (Bidon15) [reply Makefile:51]: Correct. It runs tests not under test-swamp

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:57]: This method doesn't return anything. Still thank you a lot for the suggestion as the readability is x100k better then mine 🙏

File: service/header/service.go
Comment (Bidon15) [reply service/header/service.go]: I've left the TODO with the issue 
Is it ok?

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:95]: Done

**Review by adlerjohn — COMMENTED** [review]
LGTM, will defer to others

**Review by renaynay — DISMISSED** [review]
LetsGetThisMerged

File: core/testing.go
Comment (liamsi) [inline core/testing.go:14]: Should this be const?

File: service/header/service.go
Comment (liamsi) [inline service/header/service.go:48]: Does this belong to the `IsSyncing` method?

**Review by liamsi — APPROVED** [review]
Awesome!

File: service/header/service.go
Comment (Bidon15) [reply service/header/service.go:48]: We have decided to make it a dedicated subscription API that is related to sync but not to IsSyncing itself

File: core/testing.go
Comment (Bidon15) [reply core/testing.go:14]: Ooops. Agree. Changing 
```suggestion
const defaultRetainBlocks int64 = 10
```

**Review by renaynay — DISMISSED** [review]

**Review by adlerjohn — APPROVED** [review]

**Comment by Wondertan** [comment]
Give me pls some time to make a last glance before merging.

**Comment by Bidon15** [comment]
> Give me pls some time to make a last glance before merging.

I thought that merging is too early without you comments, so sure. Take your time 💪

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:25]: It done, so remove

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:57]: nit: each field on own line

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:55]: nit: you can put it directly in &Swamp

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:65]: Blocking operations should rcv a context on them. Then pass to both subscribe and unsubscribe. Lastly, make a select inside for with ctx and <-blocks

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:119]: ?

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:168]: You don't Stop Nodes in tests and here, but we need to be nice and close them to test the whole lifecycle as well. Doing t.Cleanup calling Stop of each created Node type would be sufficient. This way we also remove the need for tests to call Stop manually each time. We can also consider returning only Started nodes on NewAnyNode

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:150]: NewLightNode for consitency.

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: Makefile
Comment (Bidon15) [reply Makefile:64]: Ooops. Fixed

File: Makefile
Comment (Bidon15) [reply Makefile:61]: Got it

File: Makefile
Comment (Bidon15) [reply Makefile:63]: Taking into account your pov, I'll leave `test` as running everything with/without data race (will be around 1-2 mins or exeuction) 
In addition, I'm renaming the unit to be explicitly `test-unit` instead of previous `test` and `test-race`

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:25]: Agreed

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:57]: Changed

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:55]: True

File: Makefile
Comment (Wondertan) [reply Makefile:63]: Awesome!

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:119]: wdym? 
The description is down below this todo

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go]: Ok

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:188]: Thanks for pointing this. Done

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:99]: Hey @vgonkivs! 
Thanks for the suggestion 🎊  . It's indeed more neat idea that Loop: 
Submitting the change

**Review by renaynay — APPROVED** [review]

File: node/tests/swamp/swamp.go
Comment (Wondertan) [reply node/tests/swamp/swamp.go:140]: Ok.

File: Makefile
Comment (Wondertan) [inline Makefile:77]: Wanted to say no to this, but later realized it might be useful to understand/see the diff that tests are failing due to races only.

File: core/testing.go
Comment (Wondertan) [inline core/testing.go:45]: optional nit: you can remove requirements of the retainBlocks param and anyway change it on the returned value directly.

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:69]: optional nit: a bit weird to pass a 2d array while it can one only array.

File: node/tests/swamp/swamp.go
Comment (Wondertan) [inline node/tests/swamp/swamp.go:133]: Respect context, pls

File: .github/workflows/test.yml
Comment (Wondertan) [inline .github/workflows/test.yml:39]: Also with a race detector?

File: Makefile
Comment (Wondertan) [reply Makefile:77]: So this is automatically resolved.

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go]: Done

File: .github/workflows/test.yml
Comment (Bidon15) [reply .github/workflows/test.yml:39]: Oops. Added

File: Makefile
Comment (Bidon15) [reply Makefile:77]: Yeap. It is indeed useful for rn. Regarding execution time - it's relatively fast, so we can consider rethinking this part later on as tests start to grow in numbers.

File: node/tests/swamp/swamp.go
Comment (Bidon15) [reply node/tests/swamp/swamp.go:71]: Well, there were 2 ideas in my head while opting for the one you are seeing here. 
1. If making this func without a 2d array, then there can be a repeated call of `stopAllNodes` in the cleanup, which is not what I wanted to see. 
2. This bit weird impl just simplifies adding future node types without the repeated calls

File: core/testing.go
Comment (Wondertan) [reply core/testing.go:45]: Ok

File: node/tests/swamp/swamp.go
Comment (Wondertan) [reply node/tests/swamp/swamp.go:71]: Ok, let's keep this.

**Review by Wondertan — APPROVED** [review]

**Comment by Wondertan** [comment]
Awesome! Huge 'thank you' for this.

**Comment by Bidon15** [comment]
> Awesome! Huge 'thank you' for this.

Arigato gozaimasu Wondertan-san 🙏

File: core/testing.go
Comment (liamsi) [inline core/testing.go:41]: I don't really see a reason to pass in the `retainBlocks` param here when everything we do is using a package-globally accessible constant here but it's ok.

File: node/tests/swamp/swamp.go
Comment (liamsi) [inline node/tests/swamp/swamp.go:52]: Are these 200ms a good choice in all scenarios?

File: node/tests/swamp/swamp.go
Comment (liamsi) [reply node/tests/swamp/swamp.go:71]: I don't understand why this could not just be a 1d array either 🤔

**Review by liamsi — APPROVED** [review]

File: core/testing.go
Comment (Bidon15) [reply core/testing.go:41]: You are right. The only motivation behind this func is to reduce the redundant code

**Pull request merged**
